internal/controllers/ipaddressclaim.go

```diff
@@ -182,6 +198,68 @@ func (r *IPAddressClaimReconciler) reconcile(ctx context.Context, claim *ipamv1.
 	return ctrl.Result{}, nil
 }
 
+func (r *IPAddressClaimReconciler) patchAddressOwnerRefsAndFinalizer(ctx context.Context, address *ipamv1.IPAddress, claim *ipamv1.IPAddressClaim, pool genericInClusterPool) error {
+	patchHelper, err := patch.NewHelper(address, r.Client)
+	if err != nil {
+		return errors.Wrap(err, "failed to initialize patch helper")
+	}
+
+	if err := r.ensureOwnerRefsAndFinalizer(ctx, address, claim, pool); err != nil {
+		return errors.Wrap(err, "failed to ensure owner refs and finalizer on address")
+	}
+
+	if err := patchHelper.Patch(ctx, address); err != nil {
+		return errors.Wrap(err, "failed to patch address")
+	}
+	return nil
+}
+
+func (r *IPAddressClaimReconciler) createAddress(ctx context.Context, claim *ipamv1.IPAddressClaim, pool genericInClusterPool, addressesInUse []ipamv1.IPAddress) (*ipamv1.IPAddress, error) {
+	address, err := r.allocateAddress(ctx, claim, pool, addressesInUse)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to allocate address")
+	}
```

Comment:
`allocateAddress` still does the job of setting the owner Refs. The call to `address := ipamutil.NewIPAddress(claim, pool)` in the function still sets the owner reference on the `IPAddress` object.

Ideally,
1. Fetch the IP address for the claim
2. If not found,  `allocateAddress` => **just** allocates a new unused IP address (does not make any API calls to persist the object, no owner ref logic in there either)
3. Use this IP address from above to set/reset finalizer and owner refs (making it clear that the controller owner ref is being set in the controller reconcile loop)
4. Persist the object to the API server.

This logic works for either scenarios (new/existing address). All the fanciful owner ref and finalizer reconciliation logic stays in one place. 
For the purpose of logging the presence of an existing IPAddress object, add an info logger after the fetch call for the IPAddress, since that was the main concern around the refactor.

Reply:
The patch helper is not able to create an address for the create case. I think there's something I'm missing.

Reply:
I get your point now, there is a way around it https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrPatch
But am gonna stop pushing for this one now. The code does what it is supposed to, just maintaining it will be difficult in the longer term. There is an easy chance that we might forget updating either of the places in the future.

Reply:
`ipamutil.NewIPAddress` was changed to no longer set the owner references. Maybe you've missed that change below?

I like it this way.

We could consider moving `ensureOwnerRefs` (without the finalizer part) to `ipamutil` as well. The idea is to potentially move `ipamutils` to CAPI to allow reusing some logic between providers. We might be able to improve it further with a second provider (e.g. the infoblox one) and turn it into some kind of frameworks. There should be quite a few similarities between ipam providers I think.

Reply:
@srm09 Ah, we got fixated on using the patch helper code, I didn't realize there was a CreateOrPatch. We've updated the logic as you've described. The result looks nice.

@schrej We've moved the ensureAddressOwnerReferences function to the `pkg/ipam-util` package.

I think we've updated everything to everyone's preference. Let us know if we've missed something.